Refactor measure

brucefan1983 opened the pull request:
**Summary**

* Refactor the measurement part to make the code more readable.
* Also prepare for enabling multiple instances of a `compute`/`dump` within one `run`.
* This PR is pure refactoring, which **should not** change any functionality, except that using multiple instances of a `compute/dump` within a run is forbidden and will trigger error message reporting.
* Compiling errors due to the missing `triclinic` member in `Box` class for NETCDF and PLUMED packages should have been fixed.

**Future work**
* Will try to enable multiple instances for some `sompute/dump` within a run.

brucefan1983 commented:
@elindgren This is a big refactoring. Could you check the tests you created still pass?

elindgren commented:
> @elindgren This is a big refactoring. Could you check if the tests you created still pass?
> 
>     * dump_dipole
> 
>     * dump_polarizability
> 
>     * active
> 
>     * dump_observer
> 
> 
> @psn417 Could you check if `dump_shock_nemd` still work as intended?

@brucefan1983 Yes, I'll run them and check 👍

elindgren commented:
@brucefan1983 It seems the following tests fail on master. I'll run a `git bisect` later to figure out when they failed. 

<img width="1696" alt="image" src="https://github.com/user-attachments/assets/632b1c42-0294-4b75-8df3-d9c58279289a" />

brucefan1983 commented:
> @brucefan1983 It seems the following tests fail on master. I'll run a `git bisect` later to figure out when they failed. Currently checking this branch as well to see if any other tests fail.
> 
> <img alt="image" width="1696" src="https://private-user-images.githubusercontent.com/27208255/416080477-632b1c42-0294-4b75-8df3-d9c58279289a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA3MzQxMjksIm5iZiI6MTc0MDczMzgyOSwicGF0aCI6Ii8yNzIwODI1NS80MTYwODA0NzctNjMyYjFjNDItMDI5NC00Yjc1LThkZjMtZDljNTgyNzkyODlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI4VDA5MTAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3NDUyYTYyMjM3ZTZiNWI5NDVlNzY5NDhmNWUxYTQ5ZmVlZTRmZGFmZjk1YTliMTE2MzZhZmY2NzJiNTE5MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6S2ZFuoanIralf1oKI5JScueXM3AjedERY2OKBVFRF0">
> EDIT: Different tests fail on this branch compared to master, so I suspect this refactor has changed things. Will give a more up-to-date comparison once I've figured out where the tests on master changed.
> 
> <img alt="image" width="1696" src="https://private-user-images.githubusercontent.com/27208255/416083797-6b061b90-2c04-4be1-bfa6-82870d667b43.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA3MzQxMjksIm5iZiI6MTc0MDczMzgyOSwicGF0aCI6Ii8yNzIwODI1NS80MTYwODM3OTctNmIwNjFiOTAtMmMwNC00YmUxLWJmYTYtODI4NzBkNjY3YjQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI4VDA5MTAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjFmYjQzMTg3ZjBlMzRlYjM2MDIwYjQyNjI5OWQ0NGU5NDEyODdjZmYxMjc3ZTdhMTIxNjNjNTU1ZDg1OWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ccXbrkCWFd1LadVLrBwHtoWYeaTtq8EkhlzuJDrSeAA">

Thanks for the tests. Are there any bugs detected?

elindgren commented:
> > @brucefan1983 It seems the following tests fail on master. I'll run a `git bisect` later to figure out when they failed. Currently checking this branch as well to see if any other tests fail.
> > <img alt="image" width="1696" src="https://private-user-images.githubusercontent.com/27208255/416080477-632b1c42-0294-4b75-8df3-d9c58279289a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA3MzQxMjksIm5iZiI6MTc0MDczMzgyOSwicGF0aCI6Ii8yNzIwODI1NS80MTYwODA0NzctNjMyYjFjNDItMDI5NC00Yjc1LThkZjMtZDljNTgyNzkyODlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI4VDA5MTAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3NDUyYTYyMjM3ZTZiNWI5NDVlNzY5NDhmNWUxYTQ5ZmVlZTRmZGFmZjk1YTliMTE2MzZhZmY2NzJiNTE5MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6S2ZFuoanIralf1oKI5JScueXM3AjedERY2OKBVFRF0">
> > EDIT: Different tests fail on this branch compared to master, so I suspect this refactor has changed things. Will give a more up-to-date comparison once I've figured out where the tests on master changed.
> > <img alt="image" width="1696" src="https://private-user-images.githubusercontent.com/27208255/416083797-6b061b90-2c04-4be1-bfa6-82870d667b43.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA3MzQxMjksIm5iZiI6MTc0MDczMzgyOSwicGF0aCI6Ii8yNzIwODI1NS80MTYwODM3OTctNmIwNjFiOTAtMmMwNC00YmUxLWJmYTYtODI4NzBkNjY3YjQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI4VDA5MTAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjFmYjQzMTg3ZjBlMzRlYjM2MDIwYjQyNjI5OWQ0NGU5NDEyODdjZmYxMjc3ZTdhMTIxNjNjNTU1ZDg1OWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ccXbrkCWFd1LadVLrBwHtoWYeaTtq8EkhlzuJDrSeAA">
> 
> Thanks for the tests. Are there any bugs detected?

I think there might be, but I need to rerun this branch after the integration tests have been fixed on master. I've made a PR with the fixed integration tests, see #933. There I also try to add the integration tests to the CI. however, we can't add the tests yet since we need a GPU runner to run them.

elindgren commented:
@brucefan1983 Only one of the integration tests fail after the refactoring, `test_active_dump_observer_average`. 

![image](https://github.com/user-attachments/assets/101139d0-81ab-41e1-bc17-e27d28287dc0)

the test checks that dump_observer should not affect the active learning cycle. However, I do not think the test makes sense, since the first run is run using the PES for the first NEP in the input list, whilst the second one on the average PES. It makes sense that the structures encountered should be slightly different, and thus have slightly different uncertainties.

I think we can probably disregard the failing test. I have changed it to instead check with and without dump_observer.

brucefan1983 commented:
Thanks for the thorough tests. I will read through the code again and
merge. Then I can start enabling multiple instances of some dump or compute
within a run. That is the major purpose of this refactoring.

Eric Lindgren ***@***.***>于周五 28.2.202516:39写道：

> @brucefan1983 <https://github.com/brucefan1983> Only one of the
> integration tests fail after the refactoring,
> test_active_dump_observer_average.
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/101139d0-81ab-41e1-bc17-e27d28287dc0>
>
> the test checks that dump_observer should not affect the active learning
> cycle. However, I do not think the test makes sense, since the first run is
> run using the PES for the first NEP in the input list, whilst the second
> one on the average PES. It makes sense that the structures encountered
> should be slightly different, and thus have slightly different
> uncertainties.
>
> I think we can probably disregard the failing test. I have changed it to
> instead check with and without dump_observer. So no bugs as far as I can
> see.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/brucefan1983/GPUMD/pull/916#issuecomment-2690808086>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF546OPJG4XQP4EJSISBQ3T2SBYL5AVCNFSM6AAAAABXHSMVI6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJQHAYDQMBYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: elindgren]*elindgren* left a comment (brucefan1983/GPUMD#916)
> <https://github.com/brucefan1983/GPUMD/pull/916#issuecomment-2690808086>
>
> @brucefan1983 <https://github.com/brucefan1983> Only one of the
> integration tests fail after the refactoring,
> test_active_dump_observer_average.
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/101139d0-81ab-41e1-bc17-e27d28287dc0>
>
> the test checks that dump_observer should not affect the active learning
> cycle. However, I do not think the test makes sense, since the first run is
> run using the PES for the first NEP in the input list, whilst the second
> one on the average PES. It makes sense that the structures encountered
> should be slightly different, and thus have slightly different
> uncertainties.
>
> I think we can probably disregard the failing test. I have changed it to
> instead check with and without dump_observer. So no bugs as far as I can
> see.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/brucefan1983/GPUMD/pull/916#issuecomment-2690808086>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF546OPJG4XQP4EJSISBQ3T2SBYL5AVCNFSM6AAAAABXHSMVI6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJQHAYDQMBYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
